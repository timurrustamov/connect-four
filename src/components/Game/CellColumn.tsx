import * as React from 'react'
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import styled from 'styled-components';

import { makeAMove } from 'store/actions';
import { State } from 'store/state';

export type CellColumnProps = {
  columnId: number,
  children?: React.ReactNode[],
  className?: string,
  isIllegal: boolean,
  makeAMove: typeof makeAMove
}

const CellColumn = (props: CellColumnProps) => {

  const { className, children, makeAMove: play, columnId } = props;
  return (
    <div onClick={() => play(columnId)} className={className}>
      {children}
    </div>
  )
}

const StyledCellColumn = styled(CellColumn) `
  display: flex;
  flex-direction: column-reverse;
  flex: 1;
  height: 100%;
`;

export default connect(
  (state: State, ownProps: { columnId: number }) => ({
    isIllegal: !!state.illegalMoves.find(move => move.columnId === ownProps.columnId)
  }),
  (dispatch) => ({
    makeAMove: bindActionCreators(makeAMove, dispatch)
  })
)(StyledCellColumn as React.ComponentType<CellColumnProps>)