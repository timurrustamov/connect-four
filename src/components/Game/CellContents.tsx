import * as React from 'react';
import styled from 'styled-components';
import { connect } from 'react-redux';
import { State } from 'store/state';

type CellContentsProps = {
  shadowColor: string,
  backgroundColor: string,
  onHoverBackgroundColor: string,
  children?: React.ReactChild,
  className?: string,
  isIllegal?: boolean
}

const CellContents = (props: CellContentsProps) => (
  <div className={props.className}>
    {props.children}
  </div>
)
const StyledCellContents = styled(CellContents) `
  position: absolute;
  top: 0;
  bottom: 0;
  right: 0;
  left: 0;
  color: palevioletred;
  font-weight: bold;
  border-radius: 50%;
  background-color: ${props => props.backgroundColor};
  box-shadow: inset 0.5rem 0.5rem 0.1rem 0px ${props => props.shadowColor};
  &:hover {
    background-color: ${props => props.onHoverBackgroundColor};
  };
`

export default connect(
  (state: State) => ({
    backgroundColor: state.theme.boardHolesColor,
    shadowColor: state.theme.boardHolesShadowColor,
    onHoverBackgroundColor: state.theme.boardHolesOnHoverColor
  })
)(StyledCellContents);