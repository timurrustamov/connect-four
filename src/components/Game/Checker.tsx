import * as React from 'react'
import styled, { keyframes } from 'styled-components';
import { UserId, State } from 'store/state';
import { connect } from 'react-redux';

const slideFromTop = keyframes`
  100% {
    transform: translate(4%, 4%);
    opacity: 1;
  }
`
const blink = keyframes`
  50% {
    background-color: gold;
  }
`

type CheckerProps = {
  userId: UserId,
  className?: string,
  withinWinningSequence?: boolean,
  playerColors: string[]
}

const Checker = (props: CheckerProps) => (
  <div className={props.className} />
)

const StyledChecker = styled(Checker) `
  position: absolute;
  width: 95%;
  height: 95%;
  border-radius: 50%;
  z-index: 1;
  transform: ${props => props.withinWinningSequence ? 'translate(4%, 4%)' : 'translate(4%, -200px)'};
  background-color: ${props => {
    const { userId, playerColors } = props;
    if (!playerColors[userId]) {
      console.warn(`${Checker.name}: did not found player ${userId} colors in theme!`);
      return 'palevioletred';
    }
    return playerColors[userId];
  }};
  animation: ${props => props.withinWinningSequence ?
    `${blink} 2s ease-in-out infinite` : `${slideFromTop} 0.3s ease-out forwards`
  };
  opacity: ${props => props.withinWinningSequence ? '1' : '0'};
`;

export default connect(
  (state: State) => ({
    playerColors: state.theme.playerColors
  })
)(StyledChecker);