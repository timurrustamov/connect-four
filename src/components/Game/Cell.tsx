import * as React from 'react';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import styled from 'styled-components';

import { makeAMove } from '@connect-four/store/actions';
import Checker from './Checker';
import { State, UserId } from 'store/state';

type CellProps = {
  makeAMove: typeof makeAMove,
  gridIndex: number,
  value?: UserId,
  columnId: number,
  rowId: number,
  boardWidth: number,
  boardHeight: number,
  className?: string,
  isIllegal?: boolean
}

type CellContentsProps = {
  children?: React.ReactChild,
  className?: string,
  isIllegal?: boolean
}

const CellContents = (props: CellContentsProps) => (
  <div className={props.className}>
    {props.children}
  </div>
)
const StyledCellContents = styled(CellContents) `
  position: absolute;
  top: 0;
  bottom: 0;
  right: 0;
  left: 0;
  color: palevioletred;
  font-weight: bold;
  border-radius: 50%;
  background-color: transparent;
  box-shadow: inset 0.5rem 0.5rem 0.1rem 0px rgba(0,0,0,0.75);
  &:hover {
    background-color: #aaa;
  };
`

const Cell = (props: CellProps) => {

  return (
    <div onClick={() => console.log(props)} className={props.className}>
      <StyledCellContents isIllegal={props.isIllegal}>
        {typeof props.value !== 'undefined' ?
          <Checker userId={props.value} /> : undefined
        }
      </StyledCellContents>
    </div>
  )
}

export const StyledCell = styled(Cell) `
  display: inline-block;
  position: relative;
  border: 10px solid #326FC8;
  background-color: white;
  border-radius: 50%;
  &:after {
    padding-top: 100%;
    display: block;
    content: '';
  }
`;

export default connect(
  (state: State) => ({
    boardWidth: state.board.options.boardWidth,
    boardHeight: state.board.options.boardHeight
  }),
  (dispatch) => ({
    makeAMove: bindActionCreators(makeAMove, dispatch)
  })
)(StyledCell)