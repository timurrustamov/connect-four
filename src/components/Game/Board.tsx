import * as React from 'react';
import { connect } from 'react-redux';
import styled from 'styled-components';

import { State } from 'store/state';
import CellColumn from 'components/Game/CellColumn';
import Cell from 'components/Game/Cell';

export type BoardProps = {
  className?: string,
  board: State['board']
}

const Board = (props: BoardProps) => {

  const { className } = props;
  const { options, grid } = props.board;
  const { boardHeight } = options;

  return (
    <div className={className}>
      {grid
        .reduce((column: typeof grid[], cell, cellIndex) => {
          const columnId = Math.floor(cellIndex / boardHeight);
          if (!column[columnId]) {
            column[columnId] = [cell];
          } else {
            column[columnId].push(cell);
          }
          return column;
        }, [])
        .map((column, columnId) => (
          <CellColumn columnId={columnId} key={columnId}>
            {column.map((cell, cellId) => {
              const gridIndex = columnId * boardHeight + cellId;
              const rowId = cellId;
              return (
                <Cell
                  value={grid[gridIndex]}
                  key={gridIndex}
                  columnId={columnId}
                  rowId={rowId}
                  gridIndex={gridIndex}
                />
              );
            })}
          </CellColumn>
        ))}
    </div>
  )
}

const StyledBoard = styled(Board) `
  display: flex;
  flex-wrap: wrap;
  width: 100%;
  min-width: ${props => props.board.options.boardWidth * 40}px;
  overflow: hidden;
`;

export default connect(
  (state: State) => ({
    board: state.board
  })
)(StyledBoard)