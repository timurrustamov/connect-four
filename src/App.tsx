import * as React from 'react';
import { Provider } from 'react-redux';
import 'typeface-roboto';
import BackButton from './components/BackButton';
import Game from './components/Game';
import Menu from './components/Menu';
import Root from './components/Root';
import ScoreBoard from './components/ScoreBoard';
import Store from './store';

/**
 * Application main component
 */
class App extends React.Component {
  render() {
    return (
      <Provider store={Store}>
        <Root>
          <Menu />
          <BackButton />
          <ScoreBoard />
          <Game />
        </Root>
      </Provider>
    );
  }
}

export default App;
