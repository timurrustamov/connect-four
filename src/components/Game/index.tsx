import * as React from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import styled from 'styled-components';
import { makeAMove } from '../../store/actions';
import { MatchState, State } from '../../store/state';
import Board from './Board';

export type RootProps = {
  board: State['board'];
  isInGame: boolean;
  backgroundColor: string;
  shadowColor: string;
  makeAMove: typeof makeAMove;
  className?: string;
};

const Root = (props: RootProps) => {
  return (
    <div className={props.className}>
      <Board />
    </div>
  );
};

const StyledRoot = styled(Root)`
  padding: 1rem;
  background-color: ${props => props.backgroundColor};
  border-radius: 10%;
  box-shadow: 25px 17px 44px -15px ${props => props.shadowColor};
  width: 100%;
  overflow: scroll;
  max-width: 720px;
  margin: 5% auto 5% auto;
  filter: ${props => (props.isInGame ? '' : 'blur(5px)')};
  @media (max-width: 700px) {
    margin: 5%;
  }
`;

/**
 * The game board with everything included!
 */
export default connect(
  (state: State) => ({
    board: state.board,
    backgroundColor: state.theme.boardBackgroundColor,
    shadowColor: state.theme.boardBackgroundShadowColor,
    isInGame: state.match.state !== MatchState.Menu
  }),
  dispatch => ({
    makeAMove: bindActionCreators(makeAMove, dispatch)
  })
)(StyledRoot);
