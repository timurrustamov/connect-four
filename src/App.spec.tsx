import { configure, shallow } from 'enzyme';
import * as React from 'react';
import * as ReactDOM from 'react-dom';
import { Provider } from 'react-redux';
import * as renderer from 'react-test-renderer';
import * as Adapter from 'enzyme-adapter-react-16';

import App from './App';
import Game from './components/Game';

configure({ adapter: new Adapter() });

it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<App />, div);
});
it('should match snapshot', () => {
  const tree = renderer.create(<App />).toJSON();
  expect(tree).toMatchSnapshot();
});
it('should contain a Redux state provider', () => {
  const container = shallow(<App />);
  expect(container.find(Provider)).toHaveLength(1);
});
it('should contain a Game board component', () => {
  const container = shallow(<App />);
  expect(container.find(Game)).toHaveLength(1);
});
