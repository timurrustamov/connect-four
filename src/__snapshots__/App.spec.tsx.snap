// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
<div
  className="sc-cSHVUG inFGiL"
>
  <div
    className="sc-jTzLTM hiRaYu"
  >
    <div
      className="sc-jzJRlG kpQEIn sc-htoDjs biMMWm"
    >
      <div
        className="sc-fjdhpX biqhvJ"
        style={
          Object {
            "flex": 1.5,
          }
        }
      >
        <h1
          style={
            Object {
              "textAlign": "center",
            }
          }
        >
          Connect-four!
        </h1>
      </div>
      <div
        className="sc-fjdhpX biqhvJ"
        style={
          Object {
            "flex": 1.5,
          }
        }
      >
        <div
          className="sc-iwsKbI cUXIRI"
        >
          <label
            className="sc-gZMcBi iwYXDQ"
          >
            Width
          </label>
          <div
            className="sc-gqjmRU bevNeV"
          >
            <button
              aria-label="Decrement Width"
              className="sc-dnqmqq QsUkQ"
              onClick={[Function]}
            >
              ⬆
            </button>
            <span
              style={
                Object {
                  "textAlign": "center",
                }
              }
            >
              7
            </span>
            <button
              aria-label="Increment Width"
              className="sc-dnqmqq QsUkQ"
              onClick={[Function]}
            >
              ⬇
            </button>
          </div>
        </div>
        <div
          className="sc-iwsKbI cUXIRI"
        >
          <label
            className="sc-gZMcBi iwYXDQ"
          >
            Height
          </label>
          <div
            className="sc-gqjmRU bevNeV"
          >
            <button
              aria-label="Decrement Height"
              className="sc-dnqmqq QsUkQ"
              onClick={[Function]}
            >
              ⬆
            </button>
            <span
              style={
                Object {
                  "textAlign": "center",
                }
              }
            >
              6
            </span>
            <button
              aria-label="Increment Height"
              className="sc-dnqmqq QsUkQ"
              onClick={[Function]}
            >
              ⬇
            </button>
          </div>
        </div>
        <div
          className="sc-iwsKbI cUXIRI"
        >
          <label
            className="sc-gZMcBi iwYXDQ"
          >
            Players
          </label>
          <div
            className="sc-gqjmRU bevNeV"
          >
            <button
              aria-label="Decrement Players"
              className="sc-dnqmqq QsUkQ"
              onClick={[Function]}
            >
              ⬆
            </button>
            <span
              style={
                Object {
                  "textAlign": "center",
                }
              }
            >
              2
            </span>
            <button
              aria-label="Increment Players"
              className="sc-dnqmqq QsUkQ"
              onClick={[Function]}
            >
              ⬇
            </button>
          </div>
        </div>
        <div
          className="sc-iwsKbI cUXIRI"
        >
          <label
            className="sc-gZMcBi iwYXDQ"
          >
            Theme
          </label>
          <div
            className="sc-gqjmRU bevNeV"
          >
            <button
              aria-label="Decrement Theme"
              className="sc-dnqmqq QsUkQ"
              onClick={[Function]}
            >
              ⬆
            </button>
            <span
              style={
                Object {
                  "textAlign": "center",
                }
              }
            >
              1
            </span>
            <button
              aria-label="Increment Theme"
              className="sc-dnqmqq QsUkQ"
              onClick={[Function]}
            >
              ⬇
            </button>
          </div>
        </div>
      </div>
      <div
        className="sc-fjdhpX biqhvJ"
      >
        <button
          className="sc-VigVT jjBymZ"
          onClick={[Function]}
        >
          Start Game!
        </button>
      </div>
    </div>
  </div>
  <div
    className="sc-gzVnrw AqnRN"
  >
    <div
      className="sc-bZQynM ckJnDK"
    >
      <div
        className="sc-EHOje bQYtny"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        <div
          className="sc-ifAKCX fduazu"
        >
          <div
            className="sc-htpNat gKrWtC"
          />
        </div>
        <div
          className="sc-ifAKCX fduazu"
        >
          <div
            className="sc-htpNat gKrWtC"
          />
        </div>
        <div
          className="sc-ifAKCX fduazu"
        >
          <div
            className="sc-htpNat gKrWtC"
          />
        </div>
        <div
          className="sc-ifAKCX fduazu"
        >
          <div
            className="sc-htpNat gKrWtC"
          />
        </div>
        <div
          className="sc-ifAKCX fduazu"
        >
          <div
            className="sc-htpNat gKrWtC"
          />
        </div>
        <div
          className="sc-ifAKCX fduazu"
        >
          <div
            className="sc-htpNat gKrWtC"
          />
        </div>
      </div>
      <div
        className="sc-EHOje bQYtny"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        <div
          className="sc-ifAKCX fduazu"
        >
          <div
            className="sc-htpNat gKrWtC"
          />
        </div>
        <div
          className="sc-ifAKCX fduazu"
        >
          <div
            className="sc-htpNat gKrWtC"
          />
        </div>
        <div
          className="sc-ifAKCX fduazu"
        >
          <div
            className="sc-htpNat gKrWtC"
          />
        </div>
        <div
          className="sc-ifAKCX fduazu"
        >
          <div
            className="sc-htpNat gKrWtC"
          />
        </div>
        <div
          className="sc-ifAKCX fduazu"
        >
          <div
            className="sc-htpNat gKrWtC"
          />
        </div>
        <div
          className="sc-ifAKCX fduazu"
        >
          <div
            className="sc-htpNat gKrWtC"
          />
        </div>
      </div>
      <div
        className="sc-EHOje bQYtny"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        <div
          className="sc-ifAKCX fduazu"
        >
          <div
            className="sc-htpNat gKrWtC"
          />
        </div>
        <div
          className="sc-ifAKCX fduazu"
        >
          <div
            className="sc-htpNat gKrWtC"
          />
        </div>
        <div
          className="sc-ifAKCX fduazu"
        >
          <div
            className="sc-htpNat gKrWtC"
          />
        </div>
        <div
          className="sc-ifAKCX fduazu"
        >
          <div
            className="sc-htpNat gKrWtC"
          />
        </div>
        <div
          className="sc-ifAKCX fduazu"
        >
          <div
            className="sc-htpNat gKrWtC"
          />
        </div>
        <div
          className="sc-ifAKCX fduazu"
        >
          <div
            className="sc-htpNat gKrWtC"
          />
        </div>
      </div>
      <div
        className="sc-EHOje bQYtny"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        <div
          className="sc-ifAKCX fduazu"
        >
          <div
            className="sc-htpNat gKrWtC"
          />
        </div>
        <div
          className="sc-ifAKCX fduazu"
        >
          <div
            className="sc-htpNat gKrWtC"
          />
        </div>
        <div
          className="sc-ifAKCX fduazu"
        >
          <div
            className="sc-htpNat gKrWtC"
          />
        </div>
        <div
          className="sc-ifAKCX fduazu"
        >
          <div
            className="sc-htpNat gKrWtC"
          />
        </div>
        <div
          className="sc-ifAKCX fduazu"
        >
          <div
            className="sc-htpNat gKrWtC"
          />
        </div>
        <div
          className="sc-ifAKCX fduazu"
        >
          <div
            className="sc-htpNat gKrWtC"
          />
        </div>
      </div>
      <div
        className="sc-EHOje bQYtny"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        <div
          className="sc-ifAKCX fduazu"
        >
          <div
            className="sc-htpNat gKrWtC"
          />
        </div>
        <div
          className="sc-ifAKCX fduazu"
        >
          <div
            className="sc-htpNat gKrWtC"
          />
        </div>
        <div
          className="sc-ifAKCX fduazu"
        >
          <div
            className="sc-htpNat gKrWtC"
          />
        </div>
        <div
          className="sc-ifAKCX fduazu"
        >
          <div
            className="sc-htpNat gKrWtC"
          />
        </div>
        <div
          className="sc-ifAKCX fduazu"
        >
          <div
            className="sc-htpNat gKrWtC"
          />
        </div>
        <div
          className="sc-ifAKCX fduazu"
        >
          <div
            className="sc-htpNat gKrWtC"
          />
        </div>
      </div>
      <div
        className="sc-EHOje bQYtny"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        <div
          className="sc-ifAKCX fduazu"
        >
          <div
            className="sc-htpNat gKrWtC"
          />
        </div>
        <div
          className="sc-ifAKCX fduazu"
        >
          <div
            className="sc-htpNat gKrWtC"
          />
        </div>
        <div
          className="sc-ifAKCX fduazu"
        >
          <div
            className="sc-htpNat gKrWtC"
          />
        </div>
        <div
          className="sc-ifAKCX fduazu"
        >
          <div
            className="sc-htpNat gKrWtC"
          />
        </div>
        <div
          className="sc-ifAKCX fduazu"
        >
          <div
            className="sc-htpNat gKrWtC"
          />
        </div>
        <div
          className="sc-ifAKCX fduazu"
        >
          <div
            className="sc-htpNat gKrWtC"
          />
        </div>
      </div>
      <div
        className="sc-EHOje bQYtny"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onMouseLeave={[Function]}
      >
        <div
          className="sc-ifAKCX fduazu"
        >
          <div
            className="sc-htpNat gKrWtC"
          />
        </div>
        <div
          className="sc-ifAKCX fduazu"
        >
          <div
            className="sc-htpNat gKrWtC"
          />
        </div>
        <div
          className="sc-ifAKCX fduazu"
        >
          <div
            className="sc-htpNat gKrWtC"
          />
        </div>
        <div
          className="sc-ifAKCX fduazu"
        >
          <div
            className="sc-htpNat gKrWtC"
          />
        </div>
        <div
          className="sc-ifAKCX fduazu"
        >
          <div
            className="sc-htpNat gKrWtC"
          />
        </div>
        <div
          className="sc-ifAKCX fduazu"
        >
          <div
            className="sc-htpNat gKrWtC"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
